File: hackerrank/problem solving/algorithms/2 - implementation/50 - chocolate feast/chocolate_feast_procedural.cs
// Source: https://www.hackerrank.com/challenges/chocolate-feast/problem?isFullScreen=true

using System;

public class ChocolateFeast
{
    public static void Main()
    {
        int nTestCases = int.Parse(Console.ReadLine());
        List<int> output = _solveTestCases(nTestCases);
        output.ForEach(Console.WriteLine);
    }

        private static List<int> _solveTestCases(int nTestCases)
        {
            List<int> output = new List<int>(new int [nTestCases]);

            for (int i = 0; i < nTestCases; i++)
            {
                List<int> testCase = Console.ReadLine().Split().Select(int.Parse).ToList();
                int amountOfMoney = testCase.First();
                int chocolateBarCost = testCase[1];
                int nWrappersToTurnInBar = testCase.Last();
                output[i] = _howManyChocolatesCanBeEaten(amountOfMoney, chocolateBarCost, nWrappersToTurnInBar);
            }

            return output;
        }

            private static int _howManyChocolatesCanBeEaten(int amountOfMoney, int chocolateBarCost, int nWrappersToTurnInBar)
            {
                int eatenChocolates = amountOfMoney / chocolateBarCost;

                for (int availableWrappers = eatenChocolates; availableWrappers >= nWrappersToTurnInBar; )
                {
                    int chocolatesForFree = availableWrappers / nWrappersToTurnInBar;
                    availableWrappers = availableWrappers - (chocolatesForFree * nWrappersToTurnInBar) + chocolatesForFree;
                    eatenChocolates += chocolatesForFree;
                }

                return eatenChocolates;
            }
}

File: hackerrank/problem solving/algorithms/2 - implementation/50 - chocolate feast/chocolate_feast_procedural.rb
# Source: https://www.hackerrank.com/challenges/chocolate-feast/problem?isFullScreen=true

def main
    n_test_cases = gets.to_i
    output = solve_test_cases(n_test_cases)
    puts output
end

    def solve_test_cases(n_test_cases)
        output = Array.new(n_test_cases)

        for i in 0...n_test_cases
            amount_of_money, chocolate_bar_cost, n_wrappers_to_turn_in_bar = gets.split.map(&:to_i)
            output[i] = how_many_chocolates_can_be_eaten(amount_of_money, chocolate_bar_cost, n_wrappers_to_turn_in_bar)
        end

        output
    end

        def how_many_chocolates_can_be_eaten(amount_of_money, chocolate_bar_cost, n_wrappers_to_turn_in_bar)
            eaten_chocolates = (amount_of_money / chocolate_bar_cost).to_i
            available_wrappers = eaten_chocolates

            while available_wrappers >= n_wrappers_to_turn_in_bar
                chocolates_for_free = (available_wrappers / n_wrappers_to_turn_in_bar).to_i
                available_wrappers = available_wrappers - (chocolates_for_free * n_wrappers_to_turn_in_bar) + chocolates_for_free
                eaten_chocolates += chocolates_for_free
            end

            eaten_chocolates
        end

main
